packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp

```diff
@@ -0,0 +1,499 @@
+          else curr = &regQueue;
+        }
+        
+      }
+      
+      //store old vertex statuses for detecting changes
+      std::vector<IcePropGrounding_Status> old_statuses(nLocalOwned,ICEPROPGS_NONE);      
+      for(size_t i = 0; i < nLocalOwned; i ++){
+        old_statuses[i] = femvData[i].getGroundingStatus();
+      }
+      
+      femv->endFill();
+      femv->doOwnedToOwnedPlusShared(Tpetra::ADD);
+      femv->doOwnedPlusSharedToOwned(Tpetra::ADD);
+
+      //detect any change in status due to communication
+      for(size_t i = 0; i < nLocalOwned; i++){
+        if(femvData[i].getGroundingStatus() != old_statuses[i]){
+          if(femvData[i].is_art){
+            artQueue.push(i);
+          } else {
+            regQueue.push(i);
+          }
+        }
+      }
+      int local_done = regQueue.empty() && artQueue.empty();
+      //this call makes sure that if any inter-processor communication changed labels
+      //we catch the changes and keep propagating them.
+      Teuchos::reduceAll<int,int>(*comm,Teuchos::REDUCE_MIN,1, &local_done,&done);
+    }
+    
+    
+  }
+  
+  //function that exchanges labels between two nodes
+  //curr_node gives its labels to neighbor.
+  void giveLabels(IcePropVtxLabel<lno_t,gno_t>& curr_node, IcePropVtxLabel<lno_t,gno_t>& neighbor){
+    IcePropGrounding_Status curr_gs = curr_node.getGroundingStatus();
+    IcePropGrounding_Status nbor_gs = neighbor.getGroundingStatus();
+    int curr_node_gid = mapWithCopies->getGlobalElement(curr_node.id);
+    //if the neighbor is full, we don't need to pass labels
+    if(nbor_gs == ICEPROPGS_FULL) return;
+    //if the current nod is empty (shouldn't happen) we can't pass any labels
+    if(curr_gs == ICEPROPGS_NONE) return;
+    //if the current node is full (and not an articulation point), pass both labels on
+    if(curr_gs == ICEPROPGS_FULL && !curr_node.is_art){
+      neighbor.first_label = curr_node.first_label;
+      neighbor.first_sender = curr_node_gid;
+      neighbor.first_used = curr_node.first_used;
+      neighbor.second_label = curr_node.second_label;
+      neighbor.second_sender = curr_node_gid;
+      neighbor.second_used = curr_node.second_used;
+      neighbor.bcc_name = curr_node.bcc_name;
+      return;
+    } else if (curr_gs == ICEPROPGS_FULL) {
+      //if it is an articulation point, and it hasn't sent to this neighbor
+      if(neighbor.first_sender != curr_node_gid){
+        //send itself as a label
+        if(nbor_gs == ICEPROPGS_NONE){
+          neighbor.first_label = curr_node_gid;
+          neighbor.first_sender = curr_node_gid;
+          neighbor.first_used = true;
+          neighbor.bcc_name = curr_node.bcc_name;
+        } else if(nbor_gs == ICEPROPGS_HALF){
+          if(neighbor.first_label != curr_node_gid){
+            neighbor.second_label = curr_node_gid;
+            neighbor.second_sender = curr_node_gid;
+            neighbor.second_used = true;
+          }
+        }
+      }
+      return;
```

Comment:
Does the updated neighbor need to be added to a queue (as at line 266)?
If not, please document why.